### PR TITLE
#459 Filters on object inventory

### PR DESCRIPTION
**Pull request by LLGuru**
The filter editor looks like follows
image image

### CODE REVIEW

**Review by marchcat — APPROVED** [review]

**Review by akleshchev — APPROVED** [review]
Might be a good idea to add a reset when user hides/reopens build tools or changes selection.

File: indra/newview/llinventoryfilter.cpp [hunk 1]
```diff
@@ -47,7 +47,7 @@
 // linden library includes
 #include "llclipboard.h"
 #include "lltrans.h"
-
+#pragma optimize("", off)
```
[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/newview/llinventoryfilter.cpp]: Fixed, thanks!

**Review by marchcat — APPROVED** [review]

File: indra/newview/llinventorybridge.cpp [hunk 2]
```diff
@@ -3123,22 +3152,21 @@ bool move_inv_category_world_to_agent(const LLUUID& object_id,
...
     {
-        it = inventory_objects.begin();
         std::shared_ptr<LLMoveInv> move_inv(new LLMoveInv);
         move_inv->mObjectID = object_id;
         move_inv->mCategoryID = category_id;
         move_inv->mCallback = callback;
         move_inv->mUserData = user_data;
 
-        for ( ; it != end; ++it)
+        for (LLPointer<LLInventoryObject> obj : inventory_objects)
```
Comment (akleshchev) [inline indra/newview/llinventorybridge.cpp:3163]: this should have been a reference

File: indra/newview/llinventorygallery.cpp [hunk 3]
```diff
@@ -3665,11 +3666,14 @@ bool dragCategoryIntoFolder(LLUUID dest_id, LLInventoryCategory* inv_cat,
         return false;
     }
 
-    if (!inv_cat) return false; // shouldn't happen, but in case item is incorrectly parented in which case inv_cat will be NULL
+    if (!inv_cat) // shouldn't happen, but in case item is incorrectly parented in which case inv_cat will be NULL
+        return false;
+
+    if (!isAgentAvatarValid())
+        return false;
```
Comment (akleshchev) [inline indra/newview/llinventorygallery.cpp:3673]: Why does this need to check isAgentAvatarValid()?

**Comment by akleshchev** [comment]
Please do not sneak in massive changes into approved comits, original commit got lost behind cosmetics that should ahve been in a separate commit where they woln't interfer with actual work.

**Pull request merged**